### ISSUE TITLE
server stopped when closing with 'No response from heartbeat'

### ISSUE THREAD

**Issue by tmds** [sockjs/sockjs-node#185]
While using the master branch of sockjs-node, node stopped giving me a stack trace to [this line](https://github.com/sockjs/sockjs-node/blob/8c12b386d36f6683e80cb64187af0f9498135852/src/trans-websocket.coffee#L102) indicating the @session is undefined.

I saw in the code @didMessage checks @session, so perhaps such a check is also necessary in heartbeat_timeout?

While looking at the code I saw some things I did not understand programming-language wise. Can you explain 
- how the assignment of [supportsheartsbeats](https://github.com/sockjs/sockjs-node/blob/8c12b386d36f6683e80cb64187af0f9498135852/src/trans-websocket.coffee#L93) works
- where [hto_ref](https://github.com/sockjs/sockjs-node/blob/8c12b386d36f6683e80cb64187af0f9498135852/src/trans-websocket.coffee#L94-L97) comes from?

Thanks

**Comment by brycekahle** [comment]
Can you post the full stack trace? Thanks.

**Comment by tmds** [comment]
I don't have a full stack trace and was kind of what you'd expect for a timer which expired.

Trying to see what might have happened:
- [websocket closed by peer](https://github.com/sockjs/sockjs-node/blob/7489fc6bf26bcf33108d28f41432d6a2433069ff/src/trans-websocket.coffee#L60)
- [calls didAbort](https://github.com/sockjs/sockjs-node/blob/4561f500e92d3d9d8f66e64d6251c3d7340b252d/src/transport.coffee#L245)
- [calls didClose](https://github.com/sockjs/sockjs-node/blob/4561f500e92d3d9d8f66e64d6251c3d7340b252d/src/transport.coffee#L256)
- [unregisters session](https://github.com/sockjs/sockjs-node/blob/4561f500e92d3d9d8f66e64d6251c3d7340b252d/src/transport.coffee#L263)
- [session is null](https://github.com/sockjs/sockjs-node/blob/4561f500e92d3d9d8f66e64d6251c3d7340b252d/src/transport.coffee#L142)

When the timer expires: [exception thrown](https://github.com/sockjs/sockjs-node/blob/7489fc6bf26bcf33108d28f41432d6a2433069ff/src/trans-websocket.coffee#L104)

Does that look feasible?

If it is important I can try to reproduce the error to capture the exact stack trace, but I don't expect it to teach us much.

**Comment by tmds** [comment]
![sockjs-stack](https://cloud.githubusercontent.com/assets/1025424/9771342/b322e242-5735-11e5-8c10-21345a582b3f.jpg)

**Comment by macinjosh32** [comment]
We observed this issue frequently on upgrading to SockJS v0.3.16. Reverting to v0.3.15 made it go away.

**Comment by sebakerckhof** [comment]
Also seeing this. This bug was reported in september, why release a new version if you know this bug is in there?

**Comment by Bobbe** [comment]
@sebakerckhof Can you show your heartbeat_delay and disconnect_delay params?

**Comment by jardakotesovec** [comment]
I use [webpack-dev-server](https://github.com/webpack/webpack-dev-server) which uses sockjs and with 0.3.16 I also get crashes described above - time to crash is varying..

**Comment by Bobbe** [comment]
Solution that works for me is change source of lib/trans-websocket.js line 157
Now it looks like this
        if (this.session){
            return this.session.close(3000, 'No response from heartbeat');
        }

**Comment by skozin** [comment]
Seeing the same error. Here is the stack trace (`heartbeat_delay` is set to `15000`):

```js
TypeError: Cannot read property 'close' of null
    at WebSocketReceiver.heartbeat_timeout (/app/node_modules/sockjs/lib/trans-websocket.js:157:26)
    at [object Object]._onTimeout (/app/node_modules/sockjs/lib/trans-websocket.js:75:24)
    at Timer.listOnTimeout (timers.js:92:15)
```

**Comment by brycekahle** [comment]
Fixed by #179

**Comment by brycekahle** [comment]
Released in 0.3.17
